src/lustre/lustreAstDependencies.ml

```diff
@@ -20,112 +20,266 @@
     Builds a dependency graph of the lustre declarations,
     to detect circular dependencies and reject them and
     re-orders node and contract declarations to resolve
-    forward references   
+    forward references.
 
-   @author: Apoorv Ingle *)
+    Note {Types of dependency analysis}: There are three different kinds of 
+    graph dependency analysis and sorting done here.
+    1. Top level constants and type declarations and functions (starts at [mk_graph_decls]) 
+    2. Nodes and contracts (starts at [mk_graph_decls])
+    3. Sorting equations of contracts and cirular analysis of node equations 
+
+   TODO: This should module should supercede LustreDependencies when it hardens.     
+
+   @author Apoorv Ingle *)
 
 module R = Res
 module LA = LustreAst
-                          
+module LH = LustreAstHelpers
+module SI = LA.SI
+
+type 'a graph_result = ('a, Lib.position * string) result  
+
+let graph_error pos err = Error (pos, err)
+
+let (>>=) = R.(>>=)                     
+let (>>) = R.(>>)                     
+          
 module G = Graph.Make(struct
                type t = LA.ident
                let compare = Stdlib.compare
                let pp_print_t = LA.pp_print_ident
              end)
 
-module IMap = Map.Make(struct
-                  type t = LA.ident
-                  let compare i1 i2 = Stdlib.compare i1 i2
-                end)
+module IMap = struct
+  include (Map.Make(struct
+               type t = LA.ident
+               let compare i1 i2 = Stdlib.compare i1 i2
+             end))
+  let keys: 'a t -> key list = fun m -> List.map fst (bindings m)             
+end
 
-type 'a graph_result = ('a, Lib.position * string) result  
-let graph_error pos err = Error (pos, err)
-let (>>=) = R.(>>=)                     
+module IntMap = struct
+  include (Map.Make(struct
+               type t = int
+               let compare i1 i2 = Stdlib.compare i1 i2
+             end))
+  let keys: 'a t -> key list = fun m -> List.map fst (bindings m)             
+end
+            
+type id_pos_map = (Lib.position list) IMap.t
+(** stores all the positions of the occurance of an id *)
+
+type node_summary = ((int list) IntMap.t) IMap.t
+(** The node summary contains the positions of the input streams 
+    of a node that are used in their current value for each output stream. *)
```

Comment:
You should include a more detailed description of this data structure. In particular, clarify that both input and outputs are referenced by its relative position in the list of input and output parameters (as opposed to the column or other reference system), and explain why it is convenient to use the position instead of the name that would be more natural. My understanding is that the position allows you to refer to inputs and outputs uniformly in both the caller and callee contexts. Is that right? You should think in the next person who will read your code :-)

Reply:
Added an example.